game/gui/display_manager.py

```diff
@@ -0,0 +1,92 @@
+import os
+
+from constants import FPS_MAX, SCREEN_WIDTH, SCREEN_HEIGHT, WHITE, MENU_FONT
+from gui.menu import *
+
+os.environ['SDL_VIDEO_CENTERED'] = '1'  # window at center
+
+
+class DisplayManager:
+
+    def __init__(self, game):
+
+        super().__init__()
+        pygame.init()
+
+        self.UP_KEY, self.DOWN_KEY, self.START_KEY, self.BACK_KEY = False, False, False, False
+
+        self.DISPLAY_W = SCREEN_WIDTH
+        self.DISPLAY_H = SCREEN_HEIGHT
+
+        self.screen = pygame.display.set_mode((SCREEN_WIDTH, SCREEN_HEIGHT))
+        self.font_name = MENU_FONT
+
+        pygame.display.set_caption("Tank Game")
+
+        self.running = True
+        self.playing = False
+
+        # Referencing the main menu object - which won't change
+        self.main_menu = MainMenu(self)
+        self.options = OptionsMenu(self)
+        self.credits = CreditsMenu(self)
+
+        # This variable will change based on the menu
+        self.curr_menu = self.main_menu
+
+        self.game = game
+
+        self.__clock = pygame.time.Clock()
+
+    def check_events(self):
+        for event in pygame.event.get():
+            if event.type == pygame.QUIT:
+                # set the game end
+                self.game.over.set()
+                self.playing = False
+                self.running = False
+                self.curr_menu.run_display = False
+            if event.type == pygame.KEYDOWN:
+                if event.key == pygame.K_RETURN:
+                    self.START_KEY = True
+                if event.key == pygame.K_BACKSPACE:
+                    self.BACK_KEY = True
+                if event.key == pygame.K_DOWN:
+                    self.DOWN_KEY = True
+                if event.key == pygame.K_UP:
+                    self.UP_KEY = True
+
+    def reset_keys(self):
+        self.UP_KEY, self.DOWN_KEY, self.START_KEY, self.BACK_KEY = False, False, False, False
+
+    def draw_text(self, text, size, x, y):
+        font = pygame.font.Font(self.font_name, size)
+        text_surface = font.render(text, True, WHITE)
+        text_rect = text_surface.get_rect()
+        text_rect.center = (x, y)
+        self.screen.blit(text_surface, text_rect)
+
+    def run(self) -> None:
```

Comment:
Did not quite understand what is the execution entry point for this game. Usually the entry point is some kind of GUI component which than internally creates game logic and starts it up. But here Game instance seems to be a top level component which holds game logic and references to GUI components. This somewhat resembles MVC pattern but Game seems to be responsible for too many things. I would recommend to think of splitting Game class to smaller pieces. One which is responsible for game logic, login, logout, managing players etc. And another one which is responsible for interaction between GUI and game logic. The last one is a good candidate for being an entry point. In some cases this last class might become too trivial - just a proxy between GUI and logic. In that case you might want to omit this class and make one of your GUI components an entry point.

Reply:
Agreed, the display manager will work well as an entry point here, and I also agree to the fact that the Game class needs to be split up, since it has grown over the course of development to be quite large.